types/module/module.go

```diff
@@ -383,27 +383,34 @@ func (m *Manager) ExportGenesis(ctx sdk.Context, cdc codec.JSONCodec) map[string
 // ExportGenesisForModules performs export genesis functionality for modules
 func (m *Manager) ExportGenesisForModules(ctx sdk.Context, cdc codec.JSONCodec, modulesToExport []string) map[string]json.RawMessage {
 	genesisData := make(map[string]json.RawMessage)
-	if len(modulesToExport) == 0 {
-		for _, moduleName := range m.OrderExportGenesis {
-			if module, ok := m.Modules[moduleName].(HasGenesis); ok {
-				genesisData[moduleName] = module.ExportGenesis(ctx, cdc)
-			}
-		}
 
-		return genesisData
+	if len(modulesToExport) == 0 {
+		modulesToExport = m.OrderExportGenesis
 	}
 
 	// verify modules exists in app, so that we don't panic in the middle of an export
 	if err := m.checkModulesExists(modulesToExport); err != nil {
 		panic(err)
 	}
 
-	for _, moduleName := range modulesToExport {
+	channels := make([]chan json.RawMessage, len(modulesToExport))
+	modulesWithGenesis := make([]string, 0, len(modulesToExport))
+
+	for i, moduleName := range modulesToExport {
 		if module, ok := m.Modules[moduleName].(HasGenesis); ok {
-			genesisData[moduleName] = module.ExportGenesis(ctx, cdc)
+			channels[i] = make(chan json.RawMessage)
+			modulesWithGenesis = append(modulesWithGenesis, moduleName)
+
+			go func(module HasGenesis, ch chan json.RawMessage) {
```

Comment:
This introduces a race condition on ` github.com/cosmos/cosmos-sdk/store/types.(*infiniteGasMeter).ConsumeGas()`.  CI on [wasmd](https://app.circleci.com/pipelines/github/CosmWasm/wasmd/2911/workflows/ed7f19dc-853d-4c2a-aefa-b61df7d02f6a/jobs/11222) found this.

Reply:
thanks for identifying this, agree we need better tests. @Reecepbcups can you help here?